GW2EIEvtcParser/EncounterLogic/Strikes/SotO/CosmicObservatory.cs

```diff
@@ -82,78 +82,76 @@ protected override CombatReplayMap GetCombatMapInternal(ParsedEvtcLog log)
+                            // Dagda uses Demonic Blast at 90% but she will not spawn the red pushback AoE
+                            // We check if she has gained the buff to be sure that the phase has started.
+                            var phaseBuff = phaseBuffs.Where(x => x.Start >= cast.Time).FirstOrNull();
+                            if (phaseBuff is null || Math.Abs(cast.Time - phaseBuff.Value.Start) > 4000)
                             {
-                                lifespan.Item2 = demonicBlastCast.Time;
-                                break;
+                                continue;
+                            }
+                            // Hardcoded positional value, Dagda isn't in the center but the AoE is
+                            lifespan = (cast.Time, phaseBuff.Value.End);
+                            var connector = new PositionConnector(new(305.26892f, 920.6105f, -5961.992f));
+                            var circle = new CircleDecoration(800, lifespan, Colors.Red, 0.4, connector);
+                            replay.Decorations.Add(circle);
+                            break;
+                        // Planet Crash - Breakbar
+                        case PlanetCrashSkill:
+                            lifespan = (cast.Time, cast.EndTime);
+                            replay.Decorations.Add(new OverheadProgressBarDecoration(CombatReplayOverheadProgressBarMajorSizeInPixel, lifespan, Colors.Red, 0.6, Colors.Black, 0.2, [(cast.Time, 0), (cast.Time + 10000, 100)], new AgentConnector(target))
+                                .UsingRotationConnector(new AngleConnector(180)));
+                            break;
+                        // Spinning Nebula
+                        case SpinningNebulaCentral:
+                        case SpinningNebulaWithTeleport:
+                            lifespan = (cast.Time, cast.Time + cast.ActualDuration);
+                            replay.Decorations.AddWithGrowing(new CircleDecoration(300, lifespan, "rgba(0, 191, 255, 0.2)", new AgentConnector(target)), lifespan.end);
+                            break;
+                        // Shooting Stars - Green Arrow
+                        case ShootingStars:
+                            uint length = 1500;
+                            uint width = 100;
+                            int castDuration = 6700;
+                            lifespan = (cast.Time, cast.Time + castDuration);
+
+                            // The mechanic gets cancelled during the intermission phases since the CM release.
+                            // Before then, the mechanic would continue during the phase and shoot.
+                            if (log.LogData.GW2Build >= GW2Builds.DagdaNMHPChangedAndCMRelease)
+                            {
+                                foreach (CastEvent demonicBlastCast in casts.Where(x => x.SkillId == DemonicBlast))
```

Comment:
better to store the result of the Where somewhere, instead of doing it for each instance of ShootingStarrs